pallets/phala/src/fat_tokenomic.rs

```diff
@@ -0,0 +1,142 @@
+//! The Fat Contract tokenomic module
+
+pub use self::pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::mq::MessageOriginInfo;
+	use frame_support::{
+		dispatch::DispatchResult,
+		pallet_prelude::*,
+		traits::{Currency, ExistenceRequirement::*, StorageVersion},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+	use phala_types::messaging::ContractId;
+	use sp_runtime::traits::{AccountIdConversion, Zero};
+
+	type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	const PALLET_ID: PalletId = PalletId(*b"phat/tok");
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type Currency: Currency<Self::AccountId>;
+	}
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(5);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	/// Stake of user to contract
+	#[pallet::storage]
+	pub type ContractUserStakes<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::AccountId,
+		Twox64Concat,
+		ContractId,
+		BalanceOf<T>,
+		ValueQuery,
+	>;
+
+	/// Map of contracts to their total stakes received
+	#[pallet::storage]
+	pub type ContractTotalStakes<T: Config> =
+		StorageMap<_, Twox64Concat, ContractId, BalanceOf<T>, ValueQuery>;
+
+	/// Minimum allowed stake
+	#[pallet::storage]
+	pub type MinStake<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		ContractDepositChanged {
+			contract: ContractId,
+			deposit: BalanceOf<T>,
+		},
+		UserStakeChanged {
+			account: T::AccountId,
+			contract: ContractId,
+			stake: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InvalidAmountOfStake,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T: crate::mq::Config,
+		T: crate::fat::Config,
+	{
+		#[pallet::weight(0)]
+		pub fn adjust_stake(
+			origin: OriginFor<T>,
+			contract: ContractId,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let user = ensure_signed(origin)?;
+			ensure!(
+				amount == Zero::zero() || amount >= MinStake::<T>::get(),
+				Error::<T>::InvalidAmountOfStake
+			);
+
+			let mut total = ContractTotalStakes::<T>::get(&contract);
+			let orig = ContractUserStakes::<T>::get(&user, &contract);
+			if amount > orig {
```

Comment:
It did return directly. I removed the return because users may stake before the contract is deployed and the stake wouldn't take effect when the contract launches later. Users can re-stake with the same amount to resend the message to contract.

Reply:
Got it. That's a good point. Let's mention in the docstring

Reply:
> Let's mention in the docstring

Docstring added